### PR TITLE
Bowling Pull

### PR DESCRIPTION
**Pull request by GJMcGowan**
Backend of challenge complete, still need to do front-end.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        </tbody>
+      </table>
+      <br><br>
+      <h1>Final Score:</h1>
+      <br>
+      <p id='finalscore'>0</p>
+    </div>
+  </body>
+
+</html>
```
Comment (spike01) [inline index.html]: Semantic HTML!

File: public/css/bowling.css [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  border: 1px solid #999;
+  padding: 0.5rem;
+}
+
+/*td:nth-child(1) { background: hsl(150, 50%, 50%); }
+td:nth-child(2) { background: hsl(160, 60%, 50%); }
+td:nth-child(3) { background: hsl(170, 70%, 50%); }
+td:nth-child(4) { background: hsl(180, 80%, 50%); }
+td:nth-child(5) { background: hsl(190, 90%, 50%); }
+td:nth-child(6) { background: hsl(200, 99%, 50%); }*/
```
Comment (spike01) [inline public/css/bowling.css]: WIP but try to leave master clean

File: public/css/reset.css [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+blockquote:before, blockquote:after,
+q:before, q:after {
+  content: '';
+  content: none;
+}
+table {
+  border-collapse: collapse;
+  border-spacing: 0;
+}
```
Comment (spike01) [inline public/css/reset.css]: :+1: on a reset

File: spec/BowlingSpec.js [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+  });
+
+  describe('Basic Scoring', function() {
+
+    it('knows which frame it is on', function() {
+      expect(bowling.bowlingFrame).toEqual(1);
+    });
+
+    it('can read back the scores for all frames', function() {
+      array = bowling.allFramesScore();
```
Comment (spike01) [inline spec/BowlingSpec.js]: array is un-descriptive but I'm splitting hairs here

File: spec/BowlingSpec.js [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+      expect(array).toEqual([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
+    });
+
+    it('can make one bowl and have the correct array returned', function() {
+      bowling.bowl(3);
+      array = bowling.allFramesScore();
```
Comment (spike01) [inline spec/BowlingSpec.js]: This is a slightly pointless intermediate variable

File: spec/BowlingSpec.js [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+  });
+
+  describe('Frames', function() {
+    it('there should be ten frames made', function() {
+      var framesNumber = Object.keys(bowling.frames).length
+      var failures = 0
+      for (i = 1; i <= framesNumber; i++) {
+        if(bowling.frames[i] instanceof Frame){
+          continue;
```
Comment (spike01) [inline spec/BowlingSpec.js]: IF you don't know what this does, you probably shouldn't be using it

File: spec/BowlingSpec.js [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        } else {
+          failures++;
+        };
+      };
+      expect(failures).toEqual(0);
+      expect(framesNumber).toEqual(10);
+    });
```
Comment (spike01) [inline spec/BowlingSpec.js]: This is overkill - you're testing JS internals! Just check for the framesNumber, then you also only have 1 expectation

File: spec/BowlingSpec.js [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    it('should be able to bowl twice and incorporate the score', function() {
+      bowling.bowl(5);
+      bowling.bowl(4);
+      expect(bowling.finalScore()).toEqual(9);
+    });
+
+    it('should be able to move on to the next frame when there have been two rolls', function() {
+      for (i = 0; i <= 2; i++) {
+        bowling.bowl(4);
+      }
```
Comment (spike01) [inline spec/BowlingSpec.js]: Just repeat yourself here

File: spec/BowlingSpec.js [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+    it('records the correct bonus when there are sequential strikes', function(){
+      bowling.bowl(10);
+      bowling.bowl(10);
+      bowling.bowl(3);
+      bowling.bowl(3);
+      expect(bowling.bowlingFrame).toEqual(4);
+      expect(bowling.frames[1].score).toEqual(23);
+      expect(bowling.frames[2].score).toEqual(16);
+      expect(bowling.allFramesScore()).toEqual([23, 16, 6, 0, 0, 0, 0, 0, 0, 0])
```
Comment (spike01) [inline spec/BowlingSpec.js]: For the final expectation, the two above MUST be true - so you don't need them

OR

Split into further tests

File: spec/BowlingSpec.js [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+    });
+
+    it('can play a game with all strikes', function() {
+      for (i = 1; i < 13; i++) {
+        bowling.bowl(10);
+      };
+      expect( function(){ bowling.bowl(1); } ).toThrow(new Error('it is the end of the game'));
+      expect(bowling.finalScore()).toEqual(300);
+    });
+
```
Comment (spike01) [inline spec/BowlingSpec.js]: Extract a multiple roll function here - you can see that it's screaming out to be made.

Don't test for the error case every time - test it once, assume that it's gonna be working still, if you find an edge case then test that

File: spec/interfaceSpec.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+  beforeEach(function(){
+    jasmine.getFixtures().fixturesPath = '.';
+    loadFixtures('index.html');
+  });
+
+  it('displays the frame scores', function() {
+    expect('.frames').toContainText('1')
+  });
+
+  xit('bowls and updates a score', function() {
```
Comment (spike01) [inline spec/interfaceSpec.js]: Try to keep master clean

File: src/Bowling.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    this.bowls = score;
+  };
+  if(this.bowlingFrame > 1){
+    this.bonus(1);
+    this.bonus(2);
+  };
+  if(this.nextFrame()) {
+    this.bowlingFrame += 1;
+  };
+};
```
Comment (spike01) [inline src/Bowling.js]: ?! Mysterious function - I really don't know what this does on first sight, and all I wanted it to do is to bowl

File: src/Bowling.js [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+Bowling.prototype.nextFrame = function() {
+  if(this.frames[this.bowlingFrame].frameTally === 2) {
```
Comment (spike01) [inline src/Bowling.js]: In most cases you can extract a more meaningful function from a conditional

File: src/Bowling.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    return true;
+  };
+};
```
Comment (spike01) [inline src/Bowling.js]:
```javascript
Bowling.prototype.nextFrame = function() {
  return this.frames[this.bowlingFrame].frameTally === 2;
  };
};
```

Comment (spike01) [inline src/Bowling.js]: Would it make more sense for it to be called isNextFrame?

File: src/Bowling.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+
+Bowling.prototype.allFramesScore = function() {
+  var allScore = []
```
Comment (spike01) [inline src/Bowling.js]: allScores?

File: src/Bowling.js [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+  for (i = 1; i<= 10; i++) {
+    allScore.push(this.frames[i].score);
+  };
+  return allScore;
+};
+
+Bowling.prototype.finalScore = function() {
+  var newArray = this.allFramesScore().reduce(function(a, b) {
+    return a + b;
```
Comment (spike01) [inline src/Bowling.js]: Just be careful of single letter variable names

Comment (spike01) [inline src/Bowling.js]: You could have a function `add` which you pass to reduce

File: src/Bowling.js [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+  });
+  return newArray;
+};
+
+Bowling.prototype.bonus = function(minus) {
+  var lastFrame = this.frames[this.bowlingFrame - minus]
+  if(lastFrame === undefined) {
```
Comment (spike01) [inline src/Bowling.js]: ????

Comment (spike01) [inline src/Bowling.js]: Then just `return` early

File: src/interface.js [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+    buttonManage();
+    updateScores();
+  });
+
+  var buttonManage = function() {
+    // Hacky fix for buttons disappearing on last round
+    // Buttons disappear due to pinsLeft being 0, however
+    // the frame function can't be updated without breaking 
+    // everything.
+    // The 'fix' allows illegal numbers to be entered however.
```
Comment (spike01) [inline src/interface.js]: Comments when you have surprising behaviour :+1:
Surprising behaviour :disappointed:

**Comment by spike01** [comment]
:checkered_flag: Work on: clear naming, revealing intent. Otherwise great submission
